parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,351 @@
+        super().reset_metrics()
+        self.metrics['loss'] = 0.0
+        self.metrics['examples'] = 0
+        self.metrics['rank'] = 0
+
+    def report(self):
+        """Report loss and mean_rank from model's perspective."""
+        m = {}
+        examples = self.metrics['examples']
+        if examples > 0:
+            m['examples'] = examples
+            if self.metrics['loss'] > 0:
+                m['loss'] = self.metrics['loss']
+                m['mean_loss'] = self.metrics['loss'] / examples
+            if self.metrics['rank'] > 0:
+                m['mean_rank'] = self.metrics['rank'] / examples
+        for k, v in m.items():
+            # clean up: rounds to sigfigs and converts tensors to floats
+            m[k] = round_sigfigs(v, 4)
+        return m
+
+    def vectorize(self, *args, **kwargs):
+        """Override options in vectorize from parent."""
+        kwargs['add_start'] = False
+        kwargs['add_end'] = False
+        kwargs['split_lines'] = True
+        return super().vectorize(*args, **kwargs)
+
+    def _get_model_file(self, opt):
+        model_file = None
+
+        # first check load path in case we need to override paths
+        if opt.get('init_model') and os.path.isfile(opt['init_model']):
+            # check first for 'init_model' for loading model from file
+            model_file = opt['init_model']
+
+        if opt.get('model_file') and os.path.isfile(opt['model_file']):
+            # next check for 'model_file', this would override init_model
+            model_file = opt['model_file']
+
+        if model_file is not None:
+            # if we are loading a model, should load its dict too
+            if (os.path.isfile(model_file + '.dict') or
+                    opt['dict_file'] is None):
+                opt['dict_file'] = model_file + '.dict'
+
+        return model_file
+
+    def _set_fixed_candidates(self, shared):
+        if shared:
+            self.fixed_candidates = shared['fixed_candidates']
+            self.fixed_candidates_vec = shared['fixed_candidates_vec']
+        else:
+            if self.opt['fixed_candidates_path']:
+                print(f"Vectorizing fixed candidates set from "
+                    f"{self.opt['fixed_candidates_path']}")
+                with open(self.opt['fixed_candidates_path'], 'r') as f:
+                    self.fixed_candidates = [line.strip() for line in
+                        f.readlines()]
+                vectorize_func = self._vectorize_text
+                candidate_vecs = [vectorize_func(cand) for cand in
+                    self.fixed_candidates]
+                self.fixed_candidates_vec, _ = padded_tensor(candidate_vecs,
+                    use_cuda=self.use_cuda)
+            else:
+                self.fixed_candidates = None
+                self.fixed_candidates_vec = None
+
+    def _warn_once(self, flag, msg):
+        if not hasattr(self, flag):
+            setattr(self, flag, True)
+            print(msg)
```

Comment:
I vote `import warning; warning.warn` instead of print

Reply:
I think that makes sense if these really were warnings, but most of these are really more informational, logging which option you chose. When they print as warnings, they look more like errors, even though nothing went wrong.